tck/mp-jwt-embedded/src/main/java/org/apache/tomee/microprofile/jwt/cdi/ClaimBean.java

```diff
@@ -0,0 +1,361 @@
+                        claimValueWrapper.setValue(() -> {
+                            final T claimValue = getClaimValue(key);
+                            return claimValue;
+                        });
+
+                    } else if (claimValueType instanceof Class) {
+                        claimValueWrapper.setValue(() -> {
+                            final T claimValue = getClaimValue(key);
+                            return claimValue;
+                        });
+
+                    } else {
+                        throw new IllegalArgumentException("Unsupported ClaimValue type " + claimValueType.toString());
+                    }
+
+                    return (T) claimValueWrapper;
+                }
+
+                // handle Optional<T>
+                if (rawTypeClass.isAssignableFrom(Optional.class)) {
+                    return getClaimValue(key);
+                }
+
+                // handle Set<T>
+                if (rawTypeClass.isAssignableFrom(Set.class)) {
+                    return getClaimValue(key);
+                }
+
+                // handle List<T>
+                if (rawTypeClass.isAssignableFrom(List.class)) {
+                    return getClaimValue(key);
+                }
+            }
+
+        } else if (annotated.getBaseType().getTypeName().startsWith("javax.json.Json")) {
+            // handle JsonValue<T> (number, string, etc)
+            return (T) toJson(key);
+
+        } else {
+            // handle Raw types
+            return getClaimValue(key);
+        }
+
+        throw new IllegalStateException("Unhandled Claim type " + annotated.getBaseType());
+    }
+
+    public static String getClaimKey(final Claim claim) {
+        return claim.standard() == Claims.UNKNOWN ? claim.value() : claim.standard().name();
+    }
+
+    private T getClaimValue(final String name) {
+        final Bean<?> bean = bm.resolve(bm.getBeans(JsonWebToken.class));
+        if (RequestScoped.class.equals(bean.getScope())) {
+            jsonWebToken = JsonWebToken.class.cast(bm.getReference(bean, JsonWebToken.class, null));
+        }
+        if (jsonWebToken == null || !bean.getScope().equals(RequestScoped.class)) {
+            logger.warning(String.format("Can't retrieve claim %s. No active principal.", name));
+            return null;
+        }
+
+        final Optional<T> claimValue = jsonWebToken.claim(name);
+        logger.finest(String.format("Found ClaimValue=%s for name=%s", claimValue, name));
+        return claimValue.orElse(null);
+    }
+
+    private JsonValue toJson(final String name) {
+        final T claimValue = getClaimValue(name);
+        return wrapValue(claimValue);
+    }
+
+    private static final String TMP = "tmp"; // todo kill this if possible
+
+    private JsonValue wrapValue(Object value) {
+        JsonValue jsonValue = null;
+
+        if (value instanceof JsonValue) {
```

Comment:
Consider Class.class.isInstance

Reply:
In this situation what would it change?
The type is known at compile time right?
And isInstance is not null compatible so I would have to deal with the null right?

Reply:
isInstance is null aware. The following line 

`System.out.println(Object.class.isInstance(null));`

writes false on the console :)

Reply:
instanceOf and Class.isInstance is not used consistently. For example, MPJWTFilter and JWTUtil also use Class.isInstance for the cases where type is known at compile time.